apps/aecore/src/aec_peer_connection.erl

```diff
@@ -0,0 +1,865 @@
+
+accept(TcpSock, Options) ->
+    {ok, {Host, Port}} = inet:peername(TcpSock),
+    lager:debug("ZZPC: got incoming ~p from ~p", [TcpSock, {Host, Port}]),
+    case gen_server:start_link(?MODULE, [Options#{ tcp_sock => TcpSock, role => responder,
+                                                   host => inet_parse:ntoa(Host)}], []) of
+        {ok, Pid} ->
+            gen_tcp:controlling_process(TcpSock, Pid),
+            gen_server:cast(Pid, owns_tcp_socket),
+            {ok, Pid};
+        Res ->
+            Res
+    end.
+
+retry(PeerCon) ->
+    gen_server:cast(PeerCon, retry).
+
+ping(PeerId) ->
+    call(PeerId, ping).
+
+get_header_by_hash(PeerId, Hash) ->
+    call(PeerId, {get_header_by_hash, Hash}).
+
+get_header_by_height(PeerId, Hash) ->
+    call(PeerId, {get_header_by_height, Hash}).
+
+get_n_successors(PeerId, Hash, N) ->
+    call(PeerId, {get_n_successors, Hash, N}).
+
+get_block(PeerId, Hash) ->
+    call(PeerId, {get_block, Hash}).
+
+get_mempool(PeerId) ->
+    call(PeerId, get_mempool).
+
+send_tx(PeerId, Tx) ->
+    call(PeerId, {send_tx, Tx}).
+
+send_block(PeerId, Tx) ->
+    call(PeerId, {send_block, Tx}).
+
+stop(PeerCon) ->
+    gen_server:cast(PeerCon, stop).
+
+
+call(PeerCon, Call) when is_pid(PeerCon) ->
+    gen_server:call(PeerCon, Call);
+call(PeerId, Call) when is_binary(PeerId) ->
+    case aec_peers:get_connection(PeerId) of
+        {ok, PeerCon}    ->
+            case PeerCon of
+                {connected, Pid} -> call(Pid, Call);
+                _ -> {error, no_connection}
+            end;
+        Err = {error, _} ->
+            Err
+    end.
+
+
+%% -- gen_server callbacks ---------------------------------------------------
+init([Opts]) ->
+    Version = <<?PROTOCOL_VERSION:16>>,
+    Genesis = aec_chain:genesis_hash(),
+    {ok, Opts#{ version => Version, genesis => Genesis }, 0}.
+
+handle_call(ping, From, State) ->
+    send_ping(State, From, local_ping_obj(State));
+handle_call({get_header_by_hash, Hash}, From, State) ->
+    send_get_header_by_hash(State, From, Hash);
+handle_call({get_header_by_height, Hash}, From, State) ->
+    send_get_header_by_height(State, From, Hash);
+handle_call({get_n_successors, Hash, N}, From, State) ->
+    send_get_n_successors(State, From, Hash, N);
+handle_call({get_block, Hash}, From, State) ->
+    send_get_block(State, From, Hash);
+handle_call(get_mempool, From, State) ->
+    send_get_mempool(State, From);
+handle_call({send_tx, Hash}, From, State) ->
+    send_send_tx(State, From, Hash);
+handle_call({send_block, Hash}, From, State) ->
+    send_send_block(State, From, Hash).
+
+handle_cast(retry, S = #{ host := Host, port := Port, status := error }) ->
+    Self = self(),
+    lager:debug("ZZPC: connecting to ~p", [{Host, Port}]),
+    Pid = spawn(fun() -> try_connect(Self, Host, Port, 1000) end),
+    {noreply, S#{ status => {connecting, Pid} }};
+handle_cast(owns_tcp_socket, S0 = #{ role := responder, tcp_sock := TcpSock }) ->
```

Comment:
I find the other cast messages straightforward because they are verbs, but I don't understand this name... perhaps it should be change to start with a verb?

Reply:
How about `give_tcp_socket_ownership`?